### DISCUSSION TITLE
ImportError raised when testing wheels on windows and macos

### DISCUSSION THREAD

**Discussion by robomics** [pypa/cibuildwheel#1759] [Q&A] [answered]
Thanks for the very useful tool!

I am struggling to debug an ImportError that is raised when building and testing a minimal pybind11 project using cibuildwheel.

CI config: [link](https://github.com/robomics/debugging-cibuildwheel/blob/main/.github/workflows/wheels.yml)
Build log: [link](https://github.com/robomics/debugging-cibuildwheel/actions/runs/7875447766/job/21487353500)

The import error reads something like: `ModuleNotFoundError: No module named 'mymodule.mymodule'` (I am not sure where the second `mymodule` is coming from).

The project itself is setup as suggested by https://github.com/pybind/cmake_example.

The only differences being:
- setup.py is calling conan to install pybind11 before building the project
- setup.py creates the following `__init__.py` file:

```python3
from .mymodule import call_me

try:
    from importlib.metadata import version
except ModuleNotFoundError:
    from importlib_metadata import version

__version__ = version("mymodule")
```

Oddly enough the tests pass on Ubuntu (see [here](https://github.com/robomics/debugging-cibuildwheel/actions/runs/7875447766/job/21487353268)).
While debugging the error, I've setup a simple GHA workflow that installs the package using pip, then runs the tests with pytest on Ubuntu, macOS and Windows.
The workflow can be found [here](https://github.com/robomics/debugging-cibuildwheel/blob/main/.github/workflows/pip.yml).
To my surprise there the tests pass on all platforms without raising ImportErrors (see [here](https://github.com/robomics/debugging-cibuildwheel/actions/runs/7875447769/job/21487342329)).

At this point I am no longer sure whether the issue lies with cibuildwheel or with the way I am setting up the project.

Any help is appreciated!

**Comment by Czaki** [comment]
> The import error reads something like: `ModuleNotFoundError: No module named 'mymodule.mymodule'` (I am not sure where the second `mymodule` is coming from).

as package is named `mymodule` them `from .mymodule import call_me` evaluates to `from mymodule.mymodule import call_me`

On Ubuntu build, contains binary file named `mymodule`:
 
![obraz](https://github.com/pypa/cibuildwheel/assets/3826210/4b06b822-ffa8-474b-9ab3-67f81354dce2)

You may try to disable tests and download created wheel to manually inspect them.

**Reply by robomics** [reply]
Thanks for your rapid reply!
I've followed your advice and temporarily disabled tests.
The wheels look ok to me (they all have proper names and contain an `__init__.py` file and a shared library).
I have successfully installed one of the Windows wheels in a VM, and was able to import the library and call `mymodule.call_me()` without errors, so I am still unsure why the wheels lead to import errors when installed by `cibuildwheel`.

**Reply by henryiii** [reply, marked as answer]
> I am not sure where the second mymodule is coming from

You have the second one in your `__init__.py`. `.mymodule` means `mymodule` from the current module, which is `mymodule/`. So that means you expect `mymodule/mymodule.<plat>.so`.

Does your shared library sit inside a matching folder, next to the `__init__.py`?

I would highly recommend `pybind/scikit_build_example` be followed instead of `pybind/cmake_example`, which is very fragile, by the way.

**Reply by henryiii** [reply]
> `# Setuptools bug causes collision between pypy and cpython artifacts`

That's not true anymore, it was fixed, by the way.

**Reply by henryiii** [reply]
IMO it looks like it's not being installed into the proper folder. It's much better to do it the scikit-build-core way, with the 'install' CMake command. CMake fixes up the paths when you use that, and you can tell it what folder to install to. You are currently trying to ship the build folder, which is not recommended.

**Reply by robomics** [reply]
Thanks for your suggestions!
I've updated the project to use scikit-build-core and things are now running smoothly.
